tests/integration_tests/util.py

```diff
@@ -42,6 +45,130 @@ def verify_ordered_items_in_text(to_verify: list, text: str):
         index = matched.start()
 
 
+def format_found(header: str, items: list) -> str:
+    """Helper function to format assertion message"""
+
+    # do nothing, allows this formatter to be "stackable"
+    if not items:
+        return ""
+
+    # if only one error put the header and the error message on a single line
+    if 1 == len(items):
+        return f"\n{header}: {items.pop(0)}"
+
+    # otherwise make a list after header
+    else:
+        return f"\n{header}:\n\t- " + "\n\t- ".join(items)
+
+
+def verify_clean_boot(
+    c: "IntegrationInstance",
+    ignore_warnings: Optional[Union[List[str], bool]] = None,
+    ignore_errors: Optional[Union[List[str], bool]] = None,
+    require_warnings: Optional[list] = None,
+    require_errors: Optional[list] = None,
+):
+    """raise assertions if the client experienced unexpected warnings or errors
+
+    fail when an required error isn't found
+
+    This function is similar to verify_clean_log, hence the similar name.
+
+    differences from verify_clean_log:
+
+    - more expressive syntax
+    - extensible (can be easily extended for other log levels)
+    - less resource intensive (no log copying required)
+    - nice error formatting
+
+    c: test instance
+    ignored_warnings: list of expected warnings to ignore,
+        or true to ignore all
+    ignored_errors: list of expected errors to ignore, or true to ignore all
+    require_warnings: Optional[list] = None,
+    require_errors: Optional[list] = None,
+    fail_when_expected_not_found: optional list of expected errors
+    """
+    ignore_errors = ignore_errors or []
+    ignore_warnings = ignore_warnings or []
+    require_errors = require_errors or []
+    require_warnings = require_warnings or []
+    status = json.loads(c.execute("cloud-init status --format=json"))
+
+    unexpected_errors = set()
+    unexpected_warnings = set()
+
+    required_warnings_found = set()
+    required_errors_found = set()
+
+    for current_error in status["errors"]:
```

Comment:
These loops are more complex than I'd like. They might work perfectly well, but it isn't obvious to me that they do. I would find set operations easier, eg:

```
missing_errors = require_errors.subtract(status["errors"])
surplus_errors = status["errors"].subtract(require_errors).subtract(ignore_errors)
```

(or whatever would actually run.)

Reply:
I agree that this would be easier logic to follow, however set operations require exact matches rather than allowing substring matches (what the code currently does), or regex matches (what we probably want in the future). I haven't yet thought of an elegant way to accomplish this using sets.